server/src/main/java/org/opensearch/common/lucene/Lucene.java

```diff
@@ -162,6 +163,25 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
         return SegmentInfos.readLatestCommit(directory);
     }
 
+    /**
+     * A variant of {@link #readSegmentInfos(Directory)} that supports reading indices written by
+     * older major versions of Lucene. The underlying implementation is a workaround since the
+     * "expert" readLatestCommit API is currently package-private in Lucene. First, all commits in
```

Comment:
I have a couple questions that might be worth answering in this comment:

- Why is `org.apache.lucene.index.SegmentInfos.readLatestCommit(Directory, int)` package private?
- How does this implementation differ from the built-in (package-private) Lucene version?
- Is there a plan to swap this implementation out for the built-in Lucene version at some point?

Reply:
1. In a word - 🤷 The [Lucene commit that introduced it](https://github.com/apache/lucene/commit/c1ae6dc07c9a988533cbe7176bdeb49e2fca1d9c) kept it package-private, with the only public access to it through `DirectoryReader.open` which is the workaround used in this method implementation.

---

2. The expert API in Lucene fetches _only_ the latest `segments_N` file and reads the `SegmentInfos` object from it. The implementation here does the following:
* Under `DirectoryReader.listCommits`, It reads the `SegmentInfos` object for the latest `segments_N` file _as well as for all older_ `segments_N` files. These are returned as a sorted list of `IndexCommit` objects which are created from the corresponding `SegmentInfos`but do not hold a reference to them.
* The latest `IndexCommit` object is then used to open a `DirectoryReader`. Underneath, the `IndexCommit` is transformed to a `SegmentInfos` object by the same method as the expert API. This `SegmentInfos` object is the one that is finally returned from this method.
  
--- 

5. I don't have any information on when Lucene plans on changing the access modifier, so I don't have a plan 😞 though I certainly _want_ to swap this out for the Lucene version.

Reply:
I should also note that the two code paths described in point 2 above have an associated test case in LuceneTests - `testReadSegmentInfosExtendedCompatibilityBaseCase`